x/mongo/driver/operation.go

```diff
@@ -510,12 +510,18 @@ func (op Operation) Execute(ctx context.Context, scratch []byte) error {
 			serviceID:    startedInfo.serviceID,
 		}
 
-		// Check if there's enough time to perform a 90th percentile round trip before the Context
-		// deadline. If not, return ErrDeadlineWouldBeExceeded and skip the actual round trip.
-		if deadline, ok := ctx.Deadline(); ok && time.Now().Add(srvr.RTT90()).After(deadline) {
-			err = internal.WrapErrorf(ErrDeadlineWouldBeExceeded,
-				"Remaining timeout %v applied from Timeout is less than 90th percentile RTT", time.Until(deadline))
-		} else {
+		// Check if there's enough time to perform a round trip before the Context deadline. If op.Timeout
+		// is set, use the 90th percentile RTT as a treshold. Otherwise, use the minimum observed RTT.
```

Comment:
I misinterpreted the spec's description of the short-circuit behavior. We only want to use the 90th percentile RTT as a treshold if `Timeout` is set for the operation. Otherwise, we should use the existing behavior (using `MinRTT` as the treshold). The "minimum RTT used to prevent sending requests" test that I previously commented out still works with this correction in the short-circuit logic. The new tests added in GODRIVER-2405 will test that 90th percentile RTT is monitored and used to prevent sending requests _when_ `Timeout` is non-nil.

Reply:
Here is some background on the minimum RTT tracking. It was added in GODRIVER-2065 to solve the same problem for a customer that the 90th percentile RTT tracking tries to solve. The 90th percentile tracking was rejected due to concerns it may bias too much on not attempting operations that may succeed.

Gating the new short-circuit behavior on the `Timeout` option seems like a safe option. I suggest noting an open question in the design: "Can minimum RTT tracking be replaced with 90th percentile RTT tracking?".